### DISCUSSION TITLE
Question about application API's

### DISCUSSION THREAD

**Discussion by tstern-masten** [nasa/cFS#358] [Q&A] [answered]
Hello!

Still getting up to speed with cFS so apologies if this is a "remedial" question. 

My question is, is there a centralized point for application API's and specifically message ID's or are they all contained within the application tree? I want to be able to have external processes easily interface with applications and it would be nice to have some common library or source to link against. 

For reference I am currently using the OSK release so I am not fully up to date with the latest cFS releases, but I believe I read somewhere in OSK that there were future plans to centralize the message ID's and add tools for generating those sources.

Any and all help is greatly appreciated!

**Comment by jbohren-hbr** [comment, marked as answer]
@tstern-masten you've touched on a very popular use case.

In the `main` cFS distribution, Message ID / Topic ID and type definitions are currently only specified in C header files. There is a big push from the community to embrace the (CCSDS SOIS EDS standard](https://public.ccsds.org/publications/SOIS.aspx) for specifying all of these interfaces in XML which can be processed by the [NASA EDS Tooling](https://github.com/nasa/EdsLib). The approach leveraging EDS would allow you to generate interface source definitions for whatever external system you need from a single source of truth.

There's a fork of the cFS distribution which shows how the EDS files can be used and processed here: https://github.com/jphickey/cfe-eds-framework/

See the following discussions for more info:
- https://github.com/nasa/cFS/discussions/307
- https://github.com/nasa/cFS/discussions/284

cc @jphickey

**Reply by tstern-masten** [reply]
Awesome I really appreciate the feedback! I will take a look at EDS as it seems like a great solution for what I am after. 
Thank you!

**Reply by tstern-masten** [reply]
@jbohren-hbr I am curious the interaction of SOIS EDS and XTCE. From what I have seen there are some tools to convert SEDS to XTCE and they are both CCSDS managed XML schema, but what is the advantage of one over the other if one exists?

**Reply by jbohren-hbr** [reply]
There's a lot of overlap in the semantic information encoded in SEDS and XTCE files. The biggest distinction from my point of view is that SEDS describes a flat set of components, each of which may have a hierarchy of interfaces at different levels of abstraction, while XTCE is describes a hierarchy of components, each of which may have interfaces which are at the same level of abstraction. These two views of a system are more or less useful depending on the lifecycle stage and stakeholder in question.

I'd definitely like to hear @skliper and @jphickey's thoughts on the different roles of these specifications since they have more experience working with them in practice. Also do let me know if the comparison above is an appropriate distinction.

**Reply by jphickey** [reply]
XTCE is intended as a ground system database whereas SEDS focuses more on embedded systems and devices.  There is a lot of overlap between them and AFAIK SEDS borrowed a lot of its data type definition constructs from XTCE.

For example, WRT limits - SEDS would describe the absolute limits of what the device or software application is capable of handling, whereas XTCE would describe the limits of what the safe or nominal operational limits would be in the context of that spacecraft/deployment.  XTCE generally has a lot more detail, and more mission-specific information, where SEDS leans toward being generic and reusable.

**Reply by pmehradnia** [reply]
@tstern-masten I am curious what tools you have come across to convert SEDS to XTCE, and also, what general direction did you take with regard to integration of the EDS? Thanks :)

**Reply by tstern-masten** [reply]
We actually pivoted slightly in our approach. We found a open source tool called [auto-yamcs](https://github.com/WindhoverLabs/auto-yamcs) that allows us to generate XTCE from our cFS binaries formatted for our ground software. With a little effort I was able to get it to work with cFS 6.7.

**Reply by pmehradnia** [reply]
@tstern-masten Interesting. Did you have to structure the cFS binary in a particular way? Does it take in the various Application .so files as input and generate the XML at the output? I am assuming you didn't pursue CCDD?

**Reply by tstern-masten** [reply]
The auto-yamcs tool is made up of several smaller tools that parse the cFS binaries and generate XTCE. 

The first one, Juicer, you run recursively on any exe's or .so's that you want to build XTCE from. This tool generates a single SQLITE database of all the symbols in the passed in binaries. The only tricky thing here is the symbols have to be used in the code somewhere so you may have to create an <app name>_symbols.c source to actually define a symbol for juicer to parse (More detail on that here: https://github.com/WindhoverLabs/xtce_generator/issues/41).

From there you run it against the next tool which is tlm_cmd_merger that takes in that database and a YML config file defining the telemetry, commands, message ID's, etc. and generates another SQLITE database.

Finally that database is converted to an XTCE.xml with the xtce_generator tool which is comes pre-formatted for import into YAMCS.

Given that we plan to use YAMCS, this solution was perfect for our needs and I have been really pleased with the documentation and support from the Windhover Labs team.

**Reply by pmehradnia** [reply]
Thanks Tim. We will look into this - your guidance is much appreciated! 
-Payam

**Comment by skliper** [comment]
I bit of additional background on the mechanics - although apps typically do include a header that defines their message ID's, these are really just intended as the default such that it's trivial to plug-and-play with the open source framework.  For a real distribution the defaults are typically all overridden with headers that are maintained within the *_defs directory (which is "owned" by the distribution).  This approach avoids the need to "clone-and-own" or modify anything within the related open source repo.  Your *_defs directory then becomes the "one-stop-shop" for all these definitions, which in turn are frequently generated via a data management system like https://github.com/nasa/CCDD or similar utilizing EDS or other common data description techniques.  Using this technique you can create a single header that includes all the various distribution mission ID headers, or whatever grouping you need to make things easy to manage.
